### DISCUSSION TITLE
Compare PDFs

### DISCUSSION THREAD

**Discussion by pmkochie** [WinMerge/winmerge#1792] [Ideas]
I recently had the need to compare 2 PDFs. I found a website that does it fairly well https://draftable.com/compare. The site looks like it OCRs the 2 PDFs and then compares the contents. The site also compares 2 Word documents.
It would be nice to be able to compare 2 documents based on the contents after an OCR scan.

**Comment by sdottaka** [comment]
WinMerge can compare text within PDF files using plugins. Two plugins available are xdocdiffPlugin and Apache Tika plugin.

[xdocdiffPlugin]
http://freemind.s57.xrea.com/xdocdiffPlugin/en/index.html

[Apache Tika]
![image](https://user-images.githubusercontent.com/98126/230129792-7f15a9e4-ddd7-459b-88b9-38750f5a708f.png)

Additionally, PDF files can be compared as images by using the following method. In image comparison, Windows' OCR function can be used to extract text from within the image.

![image](https://user-images.githubusercontent.com/98126/230129581-bb701956-098f-4e92-9fc4-c359f91585c8.png)

![image](https://user-images.githubusercontent.com/98126/230127765-db87a691-8c14-434c-ba9a-55b60f151e0f.png)

**Reply by elesbb** [reply]
How does one install Apache Tika? It downloaded as a jar file. But winmerge says it needs a dll?

**Reply by sdottaka** [reply]
When you select the Apache Tika menu item, as you did, it should download it if it's not installed and make it available for use, but what error did you get?

**Reply by elesbb** [reply]
I see!
Thank you. I originally downloaded the 3.0.0 Jar from Apache Tika website, then was trying to install it manually. Just letting WinMerge do it worked.

I did notice it downloaded 2.6.0, any way to update to the 3.0.0 jar?

**Reply by sdottaka** [reply]
> I did notice it downloaded 2.6.0, any way to update to the 3.0.0 jar?

You need to edit C:\Program Files\WinMerge\Command\Apache-Tika\tika.bat. I think it's better to wait for the next version.

**Reply by elesbb** [reply]
Thanks! Will do

**Reply by qingrenquan** [reply]
> Thanks! Will do  谢谢你，谢谢将尽

修改tika.bat就可以使用tika3.0了，因为只是调用了一下jar位置。
java8无法使用tika3.0，不过程序默认现有java环境为java19，是可以使用tika3.0的，也可以改java.bat，选择某个java版本。但是默认java19就可以了，大于java11+即可使用tika3.0。

这是问ai的tika.bat脚本，去访问d:\jar\tika-x.jar
@echo off
setlocal EnableDelayedExpansion
set TikaVer=2.6.0
set TikaJar=tika-app-%TikaVer%.jar
set DOWNLOAD_URL=https://repo1.maven.org/maven2/org/apache/tika/tika-app/%TikaVer%/%TikaJar%
set TIKA_PATH=D:\jar\%TikaJar%
set MESSAGE='Apache Tika is not installed. Do you want to download it from %DOWNLOAD_URL%'
set TITLE='Apache Tika Plugin'
set TIKA_SHA256=fa289b58a5c1bb531ace78324625512a9448aa8472b5eb88b65988964048815a

cd "%APPDATA%\WinMerge"
if not exist "D:\jar\%TikaJar%" (
  cd "%~dp0..\.."
  if not exist "D:\jar\%TikaJar%" (
    mkdir "D:\jar" 2> NUL
    cd "D:\jar"
    for %%i in (%TIKA_PATH%) do mkdir %%~pi 2> NUL
    powershell "if ((New-Object -com WScript.Shell).Popup(%MESSAGE%,0,%TITLE%,1) -ne 1) { throw }" > NUL
    if errorlevel 1 (
      echo "download is canceled" 1>&2
    ) else (
      start "Downloading..." /WAIT powershell -command "[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12; Invoke-WebRequest -Uri %DOWNLOAD_URL% -UseBasicParsing -Outfile %TIKA_PATH%"
      powershell -command "$(CertUtil -hashfile '%TIKA_PATH%' SHA256)[1] -replace ' ','' -eq '%TIKA_SHA256%'" | findstr True > NUL
      if errorlevel 1 (
        echo %TIKA_PATH%: download failed 1>&2
        del %TIKA_PATH% 2> NUL
      )
    )
  )
)

"%~dp0..\Java\java.bat" -jar "D:\jar\%TikaJar%" %3 %4 %5 %6 %7 %8 %9 "%~1" > "%~2"

我遇到了默认下载2.6，jar包损坏的问题。所以就修改这个tika.bat了找原因了（在下面你可以看到我的贴子留言）。


![image](https://github.com/user-attachments/assets/64a87912-4746-4fad-a4b0-813978d16221)
![image](https://github.com/user-attachments/assets/2dfea268-52d9-4f68-83a1-11db873c3ee8)

**Reply by qingrenquan** [reply]
> Thanks! Will do  谢谢你，谢谢将尽

请注意，如果epub带有图片，tika3.0默认会有一个ocr图片过程，winmerge可能会卡死崩溃。

可以修改tika.bat脚本，去禁用默认的ocr功能。
✅ 解决方案：禁用 Tesseract OCR
方法 1：使用环境变量临时禁用 OCR
在 PowerShell 运行：
powershell复制编辑$env:TESSDATA_PREFIX="DISABLED"
java -jar d:\jar\tika-app-3.1.0.jar -t 2.epub > extracted_text1.txt

默认的ocr效果是不好的，我执行没有ocr，而是一堆乱码其他东西（图片ocr），似乎没有正确配置tesseract ocr。
使用ocr，需要单独在java项目里去用这个tesseract ocr这个识别图片文字能力。

我现在可以正常从epub通过tika提取文本（包括图片名称，但是是文本非图片），默认tika -extract 是只提取图片。

不过由于epub是有层级的zip，其实完全可以用python改名为zip或者解压为文件夹，然后直接对比文件夹就可以

**Comment by pmkochie** [comment]
I can't download it because my company firewall blockers content from
Russia and China.

On Wed, Apr 5, 2023 at 8:33 AM Takashi Sawanaka ***@***.***>
wrote:

> WinMerge can compare text within PDF files using plugins. Two plugins
> available are xdocdiffPlugin and Apache Tika plugin.
>
> [xdocdiffPlugin]
> http://freemind.s57.xrea.com/xdocdiffPlugin/en/index.html
>
> [Apache Tika]
> [image: image]
> <https://user-images.githubusercontent.com/98126/230129792-7f15a9e4-ddd7-459b-88b9-38750f5a708f.png>
>
> Additionally, PDF files can be compared as images by using the following
> method. In image comparison, Windows' OCR function can be used to extract
> text from within the image.
>
> [image: image]
> <https://user-images.githubusercontent.com/98126/230129581-bb701956-098f-4e92-9fc4-c359f91585c8.png>
>
> [image: image]
> <https://user-images.githubusercontent.com/98126/230127765-db87a691-8c14-434c-ba9a-55b60f151e0f.png>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/WinMerge/winmerge/discussions/1792#discussioncomment-5533998>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGVYPKI4BHVQ2TCTZSTRCZ3W7WGE3ANCNFSM6AAAAAAWUFTAUU>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
--
Paul Kochie
***@***.***

 ________oooo
|___|__|__q\___\,____
L.---.\__|__|.---.(OlllllllO)
 (o)_)         (o)_)---(o)_)
